src/algebra/lattice_ordered_group.lean

```diff
@@ -0,0 +1,607 @@
+
+/-!
+# Lattice ordered groups
+
+Lattice ordered groups were introduced by [Birkhoff][birkhoff1942].
+They form the algebraic underpinnings of vector lattices, Banach lattices, AL-space, AM-space etc.
+
+This file develops the basic theory, concentrating on the commutative case written additively.
+
+## Main definitions
+
+* `lattice_ordered_group` - a (non-commutative) partially ordered (multiplicative) group which forms
+  a lattice with respect to the partial order.
+* `lattice_ordered_add_group` - the additive form of a `lattice_ordered_group`
+* `lattice_ordered_comm_group` - a lattice ordered (multiplicative) group with commutative group
+  operation
+* `lattice_ordered_add_comm_group` - the additice form of a `lattice_ordered_comm_group`
+
+## Main statements
+
+- `pos_sub_neg`: Every element `a` of a `lattice_ordered_add_comm_group` has a decomposition `a⁺-a⁻`
+   into the difference of the positive and negative component.
+- `pos_meet_neg_eq_zero`: The positive and negative components are coprime.
+- `abs_triangle`: The absolute value operation satisfies the triangle inequality.
+
+It is shown that the meet and join operations are related to the absolute value operation by a
+number of equations and inequalities.
+
+## Notations
+
+- `a⁺ = a ⊔ 0`: The *positive component* of an element `a` of a `lattice_ordered_add_comm_group`
+- `a⁻ = (-a) ⊔ 0`: The *negative component* of an element `a` of a `lattice_ordered_add_comm_group`
+* `|a| = a⊔(-a)`: The *absolute value* of an element `a` of a `lattice_ordered_add_comm_group`
+
+## Implementation notes
+
+We define `lattice_ordered_group` as extending `group` and `lattice` and
+`lattice_ordered_comm_group` as extending `comm_group` and `lattice`. Equivalent additive classes
+are defined. We then show that a `lattice_ordered_comm_group` is a `lattice_ordered_group` and an
+`ordered_comm_group`.
+
+The remainder of the file establishes basic properties of `lattice_ordered_add_comm_group`. A number
+of these results also hold in the non-commutative case ([Birkhoff][birkhoff1942],
+[Fuchs][fuchs1963]) but we have not developed that here. We have concentrated on groups written
+additively, since we are primarily interested in vector lattices.
+
+## References
+
+* [Birkhoff, Lattice-ordered Groups][birkhoff1942]
+* [Bourbaki, Algebra II][bourbaki1981]
+* [Fuchs, Partially Ordered Algebraic Systems][fuchs1963]
+* [Zaanen, Lectures on "Riesz Spaces"][zaanen1966]
+* [Banasiak, Banach Lattices in Applications][banasiak]
+
+## Tags
+
+lattice, ordered, group
+-/
+
+universe u
+
+/--
+A group `(α,*)`, equipped with a partial order `≤`, making `α` into a lattice, such that, given
+elements `a` and `b` of type `α` satisfying `a ≤ b`, for all elements `c` and `d` of `type α`,
+$$c * a * d ≤ c * b * d,$$
+is said to be a *lattice ordered group*.
+-/
+class lattice_ordered_group (α : Type u)
+  extends group α, lattice α :=
+(mul_le_mul : ∀ a b : α, a ≤ b → ∀ c d : α, c * a * d ≤ c * b * d)
```

Comment:
My understanding is that this class should not exist, and
```lean
[lattice_ordered_group α]
```
should be replaced with
```lean
[lattice α] [group α] [covariant_class α α (*) (≤)] [covariant_class α α (function.swap (*)) (≤)]
```
In case you're unconvinced they're equivalent:
```lean
example (α : Type u) :
  (Σ' [group α] [lattice α] , by exactI
    Σ' [covariant_class α α (*) (≤)] [covariant_class α α (function.swap (*)) (≤)], unit) ≃
  lattice_ordered_group α :=
{ to_fun := λ x, let g := x.1, l := x.2.1, c1 := x.2.2.1, c2 := x.2.2.2.1 in by exactI
    { mul_le_mul := λ a b hab c d, mul_le_mul' (mul_le_mul' le_rfl hab) le_rfl,
      to_lattice := l,
      to_group := g },
  inv_fun := λ g, by exactI
    ⟨_, _,
      ⟨λ a b c hbc, by simpa using lattice_ordered_group.mul_le_mul b c hbc _ 1⟩,
      ⟨λ a b c hbc, by simpa using lattice_ordered_group.mul_le_mul b c hbc 1 _⟩, ()⟩,
  left_inv := λ ⟨l, g, c1, c2, _⟩, by { dsimp, ext, simp, congr },
  right_inv := λ x, by { cases x, dsimp, congr }, }
  ```

Reply:
@adomani, do you agree with this statement? Or should the typeclass exist, but all the lemmas be about `covariant_class` instead?

Reply:
@eric-wieser Thanks - I wasn't aware of `covariant_class` before, but it does seem intended for this sort of situation.  I've pushed a commit using this approach.

The main downside I see is that it's fairly easy for the casual reader to guess what `lattice_ordered_group  α` represents mathematically, but it's somewhat less obvious what `[lattice α] [comm_group α] [covariant_class α α (*) (≤)]  [covariant_class α α (function.swap (*)) (≤)]` might mean. Is there a way of aliasing the latter as the former?

Reply:
What's nice is that if you have comm_group, covariant_class mul implies covariant_class of the swapped mul.

Reply:
> What's nice is that if you have comm_group, covariant_class mul implies covariant_class of the swapped mul.

So, for the commutative case I can just do
```
variables {α : Type u} [lattice α] [comm_group α] [covariant_class α α (*) (≤)]
```
and drop the `[covariant_class α α (function.swap (*)) (≤)]`?

Reply:
I've made this change.